Upgrade solvation thermo GAV

yunsiechung opened the pull request:
_**Note that this pull request requires simultaneous updates of RMG-Py and RMG-database. The pull request on RMG-database solvation_thermo_GAV branch is https://github.com/ReactionMechanismGenerator/RMG-database/pull/471. The last commit needs to be removed before merging this branch.**_

### Motivation or Problem
The previous group additivity (GAV) method for solvation thermo uses very limited number of groups. This pull request will upgrade the solvation thermo GAV method that uses a lot more groups.

### Description of Changes
The major changes include:

1. The solvation GAV now follows an approach similar to the gas-phase thermo GAV. It uses `group`, `ring`, `polycyclic`, `longDistanceInteraction_cyclic`, `longDistanceInteraction_noncyclic`, `halogen`, and `radical` corrections to estimate solute parameters (E, S, A, B, L). The only difference between the gas-phase thermo GAV and solvation GAV methods is that the solvation GAV separates the halogen components from the main `group.py` and applies separate halogen corrections using an approach that is similar to the radical HBI corrections: all halogen atoms are first replaced by hydrogen atoms, then the group additivity estimate is made on the replaced structure, and halogen corrections are lastly added for each halogen atom to get the final prediction. This unique approach is employed because it allows one to use the experimental solute parameter data of a replaced compound and simply apply halogen corrections to get more accurate estimates for a halogenated compound.
If a compound is both radical and halogenated, RMG will first saturate the compound and check whether the saturated structure 
can be found in the solute library. If not, it will then replace halogen atoms with hydrogens and check whether the saturated, replaced structure can be found in the solute library. The detailed description is included in the documentation as well as in the commit messages.

2. Class `Entry` now has a new attribute `data_count` for storing the number of data used to fit group values for GAV or solvent parameters. Solute groups and solvent entries loaded from RMG-dabase have `data_count` attribute with class `DataCountGAV` and `DataCountSolvent`, respectively, which store the number of data used to fit solute group values and solvent parameters. For solvent entries, the solvation energy and enthalpy mean absolute errors (MAE) associated with the fitted solvent parameters are also stored under `DataCountSolvent`.

3. A sample ipython notebook is created to show how to calculate solute parameters, solvation free energy, and solvation enthalpy using the solvation GAV. The ipython notebook also contains a code of looking up available solvents in RMG-database using a SMILES string. Currently, solvents can be only searched using names, but this notebook allows an user to look up solvents with SMILES strings. 

4. The Liquid Phase System documentations has been updated to explain all the changes.

5. Solvation groups have a few special groups that `databaseTest` cannot deal with. These special groups are skipped in `databaseTest`.


### Testing
All relevant unittests have been added. The `solvationTest.py` has been updated a lot.
The updated documentation has been checked as well by making html.

### Reviewer Tips
You can try running the ipython notebook. The new solvation GAV is very similar to the thermo gas-phase GAV.

If a reviewer thinks there are too many commits in this pull request, I can try to merge some commits together.

<!--
Checklist before submission:
 - Have you added appropriate unit tests?
 - Have you checked that all unit tests pass?
 - Is your code commented and understandable?
 - Have you updated related documentation?
 - Are the commits logically organized and informative?
 - Is your branch up to date with master?
-->

mjohnson541 commented on rmgpy/data/solvation.py:
```diff
@@ -60,7 +60,31 @@ def save_entry(f, entry):
+                    if Molecule(smiles=item.molecule[0].to_smiles()).is_isomorphic(item.molecule[0]):
+                        spc_list.append(item.molecule[0].to_smiles())
+                    else:
+                        raise DatabaseError("Not sure how to save {0!r}".format(entry.item))
```
You can fall back to adjlist if it's one species, but you can't when the solvent involves multiple species?

yunsiechung replied:
Yeah, for now, I made it possible to only save a list of solvents (multiple solvents) as a list of SMILES. We only have two types of multiple solvents in RMG-dabase, methanol + water (["CO", "O"]) and acetonitrile + water (["CC#N", "O"]), which are all stored as a list of SMILES so I thought it wold be fine.

I will also add an option of saving as a list of SMILES or adjlist for multiple solvents as well for the potential future use.

yunsiechung replied:
I change the code so that it now allows an option of falling back to adjlist for the solvent involving multiple species. (fixup commit 64cf75)
I didn't create a unittest for this because it requires writing a file, but I manually checked to make sure this works properly.

mjohnson541 commented on documentation/source/users/rmg/liquids.rst:
```diff
@@ -162,34 +162,25 @@ where `A, B, E, S` and `L` are the same solute descriptors used in the Abraham m
-#. undecane
-#. water
+The solvent descriptors (`c, e, s, a, b, l, c', a', b', e', s', l'`) are largely treated as regressed empirical coefficients.
+The solvent descriptors are fitted using in-house experimental solute parameter, solvation free energy, and solvation
```
I'm confused, did you run experiment to generate all of the data used in that paper?

yunsiechung replied:
it's fixed in this fixup commit 2d234db.

mjohnson541 commented on documentation/source/users/rmg/liquids.rst:
```diff
@@ -162,34 +162,25 @@ where `A, B, E, S` and `L` are the same solute descriptors used in the Abraham m
+or (2) `the RMG website <https://rmg.mit.edu/database/solvation/libraries/solvent/>`_. If RMG-database is installed,
+these solvent information can be found found in ``RMG-database/input/solvation/libraries/solvent.py``.
+
+You can also search for a solvent with a SMILES string by using a sample ipython notebook located at
```
"with a" => "by",  "by using" => "using"

yunsiechung replied:
it's fixed in this fixup commit 2d234db.

mjohnson541 commented on documentation/source/users/rmg/liquids.rst:
```diff
@@ -308,13 +299,90 @@ The work is in progress to implement the temperature dependent solvation free en
-O, N, and S. The value of a given descriptor for a molecule is obtained by summing the contributions from each
-fragment found in the molecule and the intercept associated with that descriptor.
+to estimate the solute descriptors (`A, B, E, L,` and `S`). RMG uses the group additivity method devised by Chung et al. [Chung2021]_
+for solute descriptor estimations. The solute descriptor group additivity follows the same scheme as the RMG's group
```
"estimations" => "estimation" throughout

yunsiechung replied:
it's fixed in this fixup commit 2d234db.

mjohnson541 commented on documentation/source/users/rmg/liquids.rst:
```diff
@@ -308,13 +299,90 @@ The work is in progress to implement the temperature dependent solvation free en
+
+If a species is a radical and contains any halogen atoms, RMG's solute descriptor group additivity method will first saturate
+the compound and check whether the saturated form can be found in the solute libraries. If yes, it uses the
+solute data of the saturated from and applies the HBI corrections on the original molecule.
```
"from" => "form"

yunsiechung replied:
it's fixed in this fixup commit 2d234db.

mjohnson541 commented on documentation/source/users/rmg/liquids.rst:
```diff
@@ -308,13 +299,90 @@ The work is in progress to implement the temperature dependent solvation free en
+solute data of the saturated from and applies the HBI corrections on the original molecule.
+If the saturated form cannot be found in the solute libraries, it then replaces the halogen atoms with hydrogen atoms
+and checks whether the saturated, replaced form can be found in the solute libraries. If yes, it uses the solute data
+of the saturated, replaces form, applies a halogen correction for each halogenated site on the saturated from, and
```
Not sure what "replaces form" means here

yunsiechung replied:
it's fixed in this fixup commit 2d234db.

mjohnson541 commented on documentation/source/users/rmg/liquids.rst:
```diff
@@ -308,13 +299,90 @@ The work is in progress to implement the temperature dependent solvation free en
+of the saturated, replaces form, applies a halogen correction for each halogenated site on the saturated from, and
+finally performs the HBI corrections on the original molecule.
+If the saturated, replaced form cannot be found in the solute libraries, it computes the solute data for the saturated,
+replaced form using the group additivity, applies a halogen correction for each halogenated site on the saturated form,
```
omit "the" in "the group additivity"

yunsiechung replied:
it's fixed in this fixup commit 2d234db.

mjohnson541 commented on documentation/source/users/rmg/liquids.rst:
```diff
@@ -308,13 +299,90 @@ The work is in progress to implement the temperature dependent solvation free en
+replaced form using the group additivity, applies a halogen correction for each halogenated site on the saturated form,
+and finally performs the HBI corrections on the original molecule.
+
+A sample ipython script that shows the solute parameter group additivity calculations and solvation free energy and
```
"and solvation" => ", solvation"

yunsiechung replied:
it's fixed in this fixup commit 2d234db.

mjohnson541 commented on documentation/source/users/rmg/liquids.rst:
```diff
@@ -372,6 +440,14 @@ and a correlation for the viscosity using parameters :math:`A, B, C, D, E`:
+or (2) `the RMG website <https://rmg.mit.edu/database/solvation/libraries/solvent/>`_. If RMG-database is installed,
+these viscosity parameters can be found found in ``RMG-database/input/solvation/libraries/solvent.py``.
+
+You can also check whether a compound of your interest has the viscosity parameters in RMG-database by using a
```
"of your interest has the" => "of interest has"

yunsiechung replied:
it's fixed in this fixup commit 2d234db.

mjohnson541 commented on ipython/estimate_solvation_thermo_and_search_available_solvents.ipynb:
```diff
@@ -0,0 +1,236 @@
+    "            if solven_data.e_h is None:\n",
+    "                print('- Solvation enthalpy calculation is not availble for this solvent.')\n",
+    "            else:\n",
+    "                print('- Solvtaion enthalpy calculation is available. The relevant Mintz solvent parameters are:')\n",
```
solvtation => solvation throughout

yunsiechung replied:
It's fixed in this fixup commit 1b7f550.

mjohnson541 reviewed (changes requested):
Looks good! Some comments/questions.

yunsiechung commented on rmgpy/data/solvation.py:
```diff
@@ -132,6 +137,87 @@ def process_old_library_entry(data):
     raise NotImplementedError()
 
 
+def add_solute_data(solute_data1, solute_data2, group_additivity=False, verbose=False):
```
I named it "add_solute_data" because there's a function called 'add_thermo_data' in 'rmgpy/data/thermo.py` that essentially does the same thing for thermo data.
I can change to "append_solute_data" if that's more clear.

yunsiechung replied:
We decided to keep the naming consistent with 'rmgpy/data/thermo.py', so I won't change the function name.

yunsiechung commented on rmgpy/data/solvation.py:
```diff
@@ -833,68 +964,481 @@ def remove_h_bonding(self, saturated_struct, added_to_radicals, added_to_pairs,
+            solute_data_sat = stable_solute_data_estimator(saturated_spec, library=self.libraries['solute'])
+            if solute_data_sat:
+                if len(solute_data_sat) != 3:
+                    raise RuntimeError("solute_data should be a tuple (solute_data, library, entry), "
```
Thanks for suggestion. I will change it.

yunsiechung replied:
I changed to ValueError in this fixup commit fe314cb.

yunsiechung commented on rmgpy/data/solvation.py:
```diff
@@ -132,6 +137,87 @@ def process_old_library_entry(data):
+    return solute_data1
+
+
+def average_solute_data(solute_data_list=None):
```
I removed the default value in the fixup commit fe314cb .

yunsiechung commented on rmgpy/data/solvation.py:
```diff
@@ -795,6 +795,24 @@ def record_ring_generic_nodes(self):
                 continue
             self.groups['ring'].generic_nodes.append(label)
 
+    def replace_halogen_with_hydrogen(self, molecule):
```
I moved this function to under class Molecule in this commit ca30579.
I didn't make it as a fixup commit because it would be fixup to two commits.

yunsiechung commented:
@mjohnson541 I made all the changes you requested. You can see my comment for each of your requested change.

If everything looks good for both this and its twin RMG-database PR  [#471](https://github.com/ReactionMechanismGenerator/RMG-database/pull/471), I can squash the fixup commits, remove the temporary commit for simultaneous updates of RMG-Py and RMG-database (be7c63d), and force push.

mjohnson541 reviewed (dismissed):
Looks ready to me, just squash and rebase!

yunsiechung commented:
@mjohnson541 I squashed and rebased. But I did not remove the temporary commit (782cb28) just to make sure this passes all the tests. 
When you merge, please remove the temporary commit (782cb28) for both this and RMG-database PR.
